esmvalcore/preprocessor/_regrid_esmpy.py

```diff
@@ -39,6 +40,130 @@
 # }
 
 
+class _ESMPyRegridder:
+    """General ESMPy regridder.
+
+    Parameters
+    ----------
+    src_cube:
+        Cube defining the source grid.
+    tgt_cube:
+        Cube defining the target grid.
+    method:
+        Regridding algorithm.
+    mask_threshold:
+        Threshold used to regrid mask of input cube.
+
+    """
+
+    def __init__(
+        self,
+        src_cube: Cube,
+        tgt_cube: Cube,
+        method: str = 'linear',
+        mask_threshold: float = 0.99,
+    ):
+        """Initialize class instance."""
+        self.src_cube = src_cube
+        self.tgt_cube = tgt_cube
+        self.method = method
+        self.mask_threshold = mask_threshold
+
+    def __call__(self, cube: Cube) -> Cube:
+        """Perform regridding.
+
+        Parameters
+        ----------
+        cube:
+            Cube to be regridded.
+
+        Returns
+        -------
+        Cube
+            Regridded cube.
+
+        """
+        src_rep, dst_rep = get_grid_representants(cube, self.tgt_cube)
+        regridder = build_regridder(
+            src_rep, dst_rep, self.method, mask_threshold=self.mask_threshold
+        )
+        result = map_slices(cube, regridder, src_rep, dst_rep)
+        return result
+
+
+class _ESMPyScheme:
+    """General irregular regridding scheme.
+
+    This class can be used in :meth:`iris.cube.Cube.regrid`.
+
+    Note
+    ----
+    See `ESMPy <http://www.earthsystemmodeling.org/
+    esmf_releases/non_public/ESMF_7_0_0/esmpy_doc/html/
+    RegridMethod.html#ESMF.api.constants.RegridMethod>`__ for more details on
+    this.
+
+    Parameters
+    ----------
+    mask_threshold:
+        Threshold used to regrid mask of source cube.
+
+    """
+
+    _METHOD = ''
+
+    def __init__(self, mask_threshold: float = 0.99):
+        """Initialize class instance."""
+        self.mask_threshold = mask_threshold
+
+    def __repr__(self) -> str:
+        """Return string representation of class."""
+        return (
+            f'{self.__class__.__name__}(mask_threshold={self.mask_threshold})'
+        )
+
+    def regridder(self, src_cube: Cube, tgt_cube: Cube) -> _ESMPyRegridder:
+        """Get regridder.
+
+        Parameters
+        ----------
+        src_cube:
+            Cube defining the source grid.
+        tgt_cube:
+            Cube defining the target grid.
+
+        Returns
+        -------
+        _ESMPyRegridder
+            Regridder instance.
+
+        """
+        return _ESMPyRegridder(
+            src_cube,
+            tgt_cube,
+            method=self._METHOD,
+            mask_threshold=self.mask_threshold,
+        )
+
+
+class ESMPyAreaWeighted(_ESMPyScheme):
```

Comment:
The `regridder` method of this class (and also for the other schemes in this file) returns an `_ESMPyRegridder` according to the documentation. It would be nice to avoid leaking private classes into the documentation, would it be possible to return something public instead? And hide the base class because it is also private? I believe there is a keyword that you can add for that.